pkg/maputil/maputil.go

```diff
@@ -60,32 +63,141 @@ func recursivelyStringifyMapKey(v interface{}) (interface{}, error) {
-	}
+type arg interface {
+	shift(map[string]interface{}, func(map[string]interface{}))
+	assign(map[string]interface{}, string)
+}
 
-	k := key[0]
+type keyArg struct {
+	key string
+}
 
-	if len(key) == 1 {
-		m[k] = value
-		return m
+func (a keyArg) ensureMap(m map[string]interface{}) map[string]interface{} {
+	nested, ok := m[a.key]
+	if !ok {
+		nested = map[string]interface{}{}
+	}
+	switch t := nested.(type) {
+	case map[string]interface{}:
+		return t
+	default:
+		panic(fmt.Errorf("unexpected type: %v(%T)", t, t))
 	}
+}
+
+func (a keyArg) shift(m map[string]interface{}, next func(map[string]interface{})) {
+	t := a.ensureMap(m)
+	next(t)
+	m[a.key] = t
+}
 
-	remain := key[1:]
+func (a keyArg) assign(m map[string]interface{}, value string) {
+	m[a.key] = value
+}
 
-	nested, ok := m[k]
+type indexedKeyArg struct {
+	key   string
+	index int
+}
+
+func (a indexedKeyArg) ensureArray(m map[string]interface{}) []interface{} {
+	nested, ok := m[a.key]
 	if !ok {
-		nested = map[string]interface{}{}
+		nested = make([]interface{}, a.index+1)
 	}
 	switch t := nested.(type) {
+	case []interface{}:
+		if len(t) <= a.index {
+			t2 := make([]interface{}, a.index+1)
+			copy(t, t2)
+			t = t2
+		}
+		return t
+	default:
+		panic(fmt.Errorf("unexpected type: %v(%T)", t, t))
+	}
+}
+
+func (a indexedKeyArg) shift(m map[string]interface{}, next func(map[string]interface{})) {
+	t := a.ensureArray(m)
+	if t[a.index] == nil {
+		t[a.index] = map[string]interface{}{}
+	}
+	switch t := t[a.index].(type) {
 	case map[string]interface{}:
-		nested = Set(t, remain, value)
+		next(t)
 	default:
 		panic(fmt.Errorf("unexpected type: %v(%T)", t, t))
 	}
+	m[a.key] = t
+}
 
-	m[k] = nested
+func (a indexedKeyArg) assign(m map[string]interface{}, value string) {
+	t := a.ensureArray(m)
+	t[a.index] = value
+	m[a.key] = t
+}
 
-	return m
+func getArg(key string) arg {
+	key = strings.ReplaceAll(key, "[", " ")
+	key = strings.ReplaceAll(key, "]", " ")
+	k := key
+	idx := 0
+
+	n, err := fmt.Sscanf(key, "%s %d", &k, &idx)
+
+	if n == 2 && err == nil {
+		return indexedKeyArg{
+			key:   k,
+			index: idx,
+		}
+	}
+
+	return keyArg{
+		key: key,
+	}
+}
+
+func ParseKey(key string) []string {
+	r := []string{}
+	part := ""
+	escaped := false
+	for _, rune := range key {
+		split := false
+		switch {
+		case escaped == false && rune == '\\':
+			escaped = true
+			continue
+		case rune == '.':
+			split = escaped == false
+		}
+		escaped = false
+		if split {
+			r = append(r, part)
+			part = ""
+		} else {
+			part = part + string(rune)
+		}
+	}
+	if len(part) > 0 {
+		r = append(r, part)
+	}
+	return r
+}
+
+func Set(m map[string]interface{}, key []string, value string) {
+	if len(key) == 0 {
+		panic(fmt.Errorf("bug: unexpected length of key: %d", len(key)))
+	}
+
+	appender := getArg(key[0])
+
+	if len(key) == 1 {
+		appender.assign(m, value)
+	} else {
+		remain := key[1:]
+		appender.shift(m, func(m map[string]interface{}) {
```

Comment:
I feel like I'd prefer:

```
nested := appender.getMap(m)
Set(nested, remain, value)
```

But please correct me this appender/assign/shift is a common idioms for writing algorithm like this 😃

Reply:
Not sure why i implemented it this way, probably because i was in a functional mood ;)

I will change that right away.

Reply:
done